core/chains/evm/txmgr/nonce_tracker.go

```diff
@@ -0,0 +1,174 @@
+		Min:    100 * time.Millisecond,
+		Max:    5 * time.Second,
+		Jitter: true,
+	}
+
+	localSequence, err := s.GetNextSequence(ctx, addr)
+	// Address not found in map so skip sync
+	if err != nil {
+		s.lggr.Criticalw("Failed to retrieve local next sequence for address", "address", addr.String(), "err", err)
+		return
+	}
+
+	// Enter loop with retries
+	var attempt int
+	for {
+		select {
+		case <-chStop:
+			return
+		case <-time.After(sequenceSyncRetryBackoff.Duration()):
+			attempt++
+			err := s.SyncOnChain(ctx, addr, localSequence)
+			if err != nil {
+				if attempt > 5 {
+					s.lggr.Criticalw("Failed to sync with on-chain sequence", "address", addr.String(), "attempt", attempt, "err", err)
+				} else {
+					s.lggr.Warnw("Failed to sync with on-chain sequence", "address", addr.String(), "attempt", attempt, "err", err)
+				}
+				continue
+			}
+			return
+		}
+	}
+}
+
+func (s *nonceTracker) SyncOnChain(ctx context.Context, addr common.Address, localSequence evmtypes.Nonce) error {
+	nonce, err := s.client.PendingSequenceAt(ctx, addr)
+	if err != nil {
+		return err
+	}
+	if nonce > localSequence {
+		s.lggr.Warnw(fmt.Sprintf("address %s has been used before, either by an external wallet or a different Chainlink node. "+
+			"Local nonce is %v but the on-chain nonce for this account was %v. "+
+			"It's possible that this node was restored from a backup. If so, transactions sent by the previous node will NOT be re-org protected and in rare cases may need to be manually bumped/resubmitted. "+
+			"Please note that using the chainlink keys with an external wallet is NOT SUPPORTED and can lead to missed or stuck transactions. ",
+			addr, localSequence, nonce),
+			"address", addr.String(), "localNonce", localSequence, "chainNonce", nonce)
+
+		s.lggr.Infow("Fast-forward sequence", "address", addr, "newNextSequence", nonce, "oldNextSequence", localSequence)
+	}
+
+	s.sequenceLock.Lock()
+	defer s.sequenceLock.Unlock()
+	s.nextSequenceMap[addr] = max(localSequence, nonce)
+	return nil
+}
+
+func (s *nonceTracker) GetNextSequence(ctx context.Context, address common.Address) (seq evmtypes.Nonce, err error) {
+	s.sequenceLock.Lock()
+	defer s.sequenceLock.Unlock()
+	// Get next sequence from map
+	seq, exists := s.nextSequenceMap[address]
+	if exists {
+		return seq, nil
+	}
+
+	s.lggr.Infow("address not found in local next sequence map. Attempting to search and populate sequence.", "address", address.String())
+	// Check if address is in the enabled address list
+	if !slices.Contains(s.enabledAddresses, address) {
+		return seq, fmt.Errorf("address disabled: %s", address)
+	}
+
+	// Try to retrieve next sequence from tx table or on-chain to load the map
+	// A scenario could exist where loading the map during startup failed (e.g. All configured RPC's are unreachable at start)
+	// The expectation is that the node does not fail startup so sequences need to be loaded during runtime
+	foundSeq, err := s.getSequenceForAddr(ctx, address)
```

Comment:
This makes an RPC call onchain, while holding the lock. So we are locking this for potentially many seconds.
Could we  acquire the lock only when we are reading/writing from the nextSequenceMap?

Reply:
Couldn't we run into an issue where multiple calls are made to `GetNextSequence` where none find the value locally so they all search on-chain? Then if we only lock the actual write to the map, the callers could get different values from on-chain depending which RPC their request was sent to. They would then write conflicting values to the map and return different nonces? I think this wouldn't be a problem in the current setup where the Broadcaster only uses the nonce tracker but if other components start using this we could maybe run into this problem. My worries could also be completely unfounded though so not against making this change if this isn't a worry.

Reply:
Yes I am assuming the broadcaster is the only caller, and for a given FromAddress, it only calls NonceTracker from a single thread.
But lets keep the logic as it is for now.